### DISCUSSION TITLE
How to set minimum/maximum widht & height of control?

### DISCUSSION THREAD

**Discussion by Intelli-Soft** [AvaloniaUI/Avalonia#7765] [Q&A] [answered]
Hey
I am new to Avalonia.
I want to shift my Winforms project.
My LOB project will have a lot of textboxes (numeric ones)
So, I want to create a textbox, which has a predefined size, based on the used font size
A very simple example:
Let's say, I have a textbox, which allows to enter the degree of an circle. We know, the maximum is 360°.
So, now the size of the used textbox should have the maximum and minimum width and height set, sothat the value of "360°" is still visible. No matter on which platform I use it.
Could someone please give me advise, how I can create such a control?
Please, I am not talking about setting the minimum/maximum value, which is allowed within the control!!!
I am talking about the size of the control

THX a lot in advance

**Comment by robloo** [comment, marked as answer]
I'll answer your question directly:

You first need to measure the size of text. That can be done using the FormattedText and then the Bounds property. You will have to know fontsize, string, etc. but it will tell you the dimensions of the text. Then with the TextBox you will need to set the width to the measured value (with additional margin for the TextBox border and padding).

That said, I don't think you should be doing this. Avalonia works like other XAML frameworks such as WPF and is conceptually different than WinForms. With WinForms layout of controls is largely absolute. You put controls on a form, set their left/top position and can then anchor them to a side, etc. so they expand. It's a simple layout system.

With Avalonia it is quite a bit more complex. Unless you are using a Canvas control (which you shouldn't be for this) layout is not absolute. Controls are composable and the UI is constructed by putting controls within other controls. But the layout system is also smart enough to expand parent controls if a child control needs more space. This is done automatically for text controls like a TextBox and should be done automatically for a NumericaUpDown (but I haven't checked it). This means the control width should automatically increase based on text length.

That said, each control also has Height/Width/MinHeight/MinWidth property you can use to constrain size. Also important are HorizontalAlignment and VerticalAlignment properties.

It sounds like you are new to XAML so this is a difficult question to answer. You have to change the way you think about UI a bit. My recommendation is to find a WPF tutorial that does something similar. Then translate that to Avalonia which should be fairly simple.

**Reply by Intelli-Soft** [reply]
Hey @robloo 

First of all, thx for your very detailed answer.
I have to be honest, that you are right, that I am relative new to XAML. But I don't agree, that I have to rethink UI techniques at all... (But, maybe I am wrong...)
So, please let me explain some points.
1) You are talking about formatted text method, which is only available for Desktop in .NET 5, but is not crossplatform available
[Microsoft Doc](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.formattedtext?view=windowsdesktop-6.0)

2) Let's explain in detail, how things could look like.
Based on the idea, I have a grid layout with two columns
First column Label, that describes, what the input should do, and the second colum holds the numeric box.
Based on my first post. - Column[0], Label Text = Degrees, Column[1] = Numeric input

How could it look nice, on different screen sizes!?
When the input box uses more than double the size as the maximum input width, it will not look good. - It will look awful...
So, I would have to fix the size of the numeric input to a maximum width (based on the used font)
But this must be measured anyhow, and it must be measured crossplatform independent.

Hope you understand, what I mean.

THX a lot in advance.

**Reply by robloo** [reply]
1. No, this has nothing to do with .NET5. [`FormattedText`](http://reference.avaloniaui.net/api/Avalonia.Media/FormattedText/) is a class Avalonia provides. However, certain features may be brand-new so you may need to use the nightly build Nugets. Again, I do not recommend doing what you are trying to do though. This is fully cross-platform.
 2. The convention is now to put the 'label' above the input TextBox. This would appear as a header. I would use the `Label` control and set the target to the input `TextBox` it corresponds to (that part is like WinForms). Then the `TextBox` will have HorizontalAlignment=Stretch within it's column -- do NOT set a width. If you do want to put them side-by-side, make the label column 0 auto width and then the textbox column 1 width * (star) to take up the remaining space. Set a minimum column width of column 1 to ensure your input TextBox is always visible.

```xaml
<Grid Grid.ColumnDefinitions="Auto,*" Grid.RowDefinitions="Auto,Auto,Auto">>
    <Label Grid.Row="0" Grid.Column="0"/>
   <NumericUpDown Grid.Row="0" Grid.Column="1" HorizontalAlignment="Stretch" />
   <!-- More rows here -->
</Grid>
```

You can set MinWidth/MaxWidth on either the column or NumericUpDown to further adjust things.

If none of what I said makes sense, you will have to do a bit of reading. You can't just jump from WinForms to WPF/Avalonia without changing the way of doing things. WPF had a pretty big learning curve when we all first came over from WinForms and the only fix is just to learn. It's more like HTML with controls inside controls than it is like WinForms.

**Reply by Intelli-Soft** [reply]
Hey @robloo 
THX a lot once again for your answer.
That there is a FormattedText method, which is directly implemented within Avalon, I didn't know. - THX for this!

To be honest, I don't understand, the problem, why it shouldn't be done.
As you said by your self: _You can set MinWidth/MaxWidth on either the column or NumericUpDown to further adjust things._
OK - This is, what I try to achieve. - But I don't want to set this fixed size.
It should be calculated, based on the expected size.

Maybe I found a perfect example to explain, what I want. Please take a look here: [Example](http://www.codearsenal.net/2012/06/wpf-textbox-validation-idataerrorinfo.html#.YiiYtImZOM8)

As you can see in this example, the salary text box is sized fixed with a width of 50. And as it can be seen, that salary has a value of 1000,
It fit's almost perfect. But what, when the salary would be 100000 (cool salary... ;) ), it would not fit - you would just see a couple of 0 , but not the leading 1 anymore.
I could fix this, when I know the maximum used salary and make it the correct size, including the trailing currency sign for example.
Or, I could also do it, by just letting it without a width, but then it would take the same space as the name text. - This would not look good (I think).

THX for the time, writing/reading

**Reply by MarkusKgit** [reply]
Maybe I don't understand you completely but that is exactly what the layout system can do for you. Look at this rather minimal sample (Buttons left out for brevity):
```xaml
<Grid ColumnDefinitions="Auto,Auto" RowDefinitions="Auto, Auto">
	<Label VerticalAlignment="Center" Content="Salary:"/>
	<TextBox VerticalAlignment="Center" Grid.Column="1" MinWidth="1" Text="{Binding Salary, Converter={StaticResource DecToCurrency}}"/>
</Grid>
```

If I change the salary in this case, the Textbox changes its size automatically. Note that I set MindWidth to 1 because from the default style it has a certain MinWidth defined.

![AutoWidth](https://user-images.githubusercontent.com/13160892/157452017-95d8afe0-d33a-442e-8228-cfba8f9fa4f6.gif)

Same example but with ´FontSize="50"´ on the TextBox:

![Fontsize](https://user-images.githubusercontent.com/13160892/157452676-154fa2fb-d5c5-49c6-bafc-67837d4f01a2.png)

**Reply by Intelli-Soft** [reply]
@MarkusKgit 
Yes - that's it! - Then I maybe missunderstood something completely from my point of view.
THX a lot!

**Reply by MarkusKgit** [reply]
Once you get the hang of it, you will love it :-) It is quite the learning curve, especially when you also want to go the MVVM route but it is worth it imo.

**Reply by robloo** [reply]
I probably should have just given you the recommended code directly rather than trying to explain everything 😃

Keep in mind only that current design convention is to put header above input. Also NumericUpDown will automatically constrain input to numbers and a specified range if you wish.
